File: algorithms/cpp/46.cpp
#include <iostream>
#include <vector>
using namespace std;

class Solution
{
private:
    void dfs(vector<int> &nums, vector< vector<int> > &result, vector<int> &per, vector<bool> &visited)
    {
        if ( per.size() == nums.size() )
        {
            result.push_back(per);
            return;
        }
        else
        {
            for ( int i = 0; i < (int)nums.size(); i++ )
            {
                if ( !visited[i] )
                {
                    visited[i] = true;
                    per.push_back(nums[i]);
                    dfs(nums, result, per, visited);
                    per.pop_back();
                    visited[i] = false;
                }
            }
        }
    }
public:
    vector< vector<int> > permute(vector<int>& nums)
    {
        vector< vector<int> > result;
        vector<int> per;
        vector<bool> visited(nums.size(), false);
        dfs(nums, result, per, visited);
        return result;
    }
};

int main()
{
    return 0;
}

File: algorithms/python3/46.py
from itertools import product

class Solution:
    def permute(self, nums):
        """
        :type nums: List[int]
        :rtype: List[List[int]]
        """
        l = len(nums)
        if l <= 1:
            return [nums]
        pers = self.permute(nums[:l-1])
        r = []
        for p in product(pers, range(l)):
            r.append(p[0][:])
            r[-1].insert(p[1], nums[-1])
        return r
